awx/main/scheduler/task_manager.py

```diff
@@ -255,72 +238,52 @@ def process_finished_workflow_jobs(self, workflow_jobs):
                 workflow_job.send_notification_templates('succeeded' if workflow_job.status == 'successful' else 'failed')
         return result
 
-    @timeit
-    def start_task(self, task, instance_group, dependent_tasks=None, instance=None):
-        self.subsystem_metrics.inc("task_manager_tasks_started", 1)
-        self.start_task_limit -= 1
-        if self.start_task_limit == 0:
-            # schedule another run immediately after this task manager
-            schedule_task_manager()
-        from awx.main.tasks.system import handle_work_error, handle_work_success
+    def timeout_approval_node(self):
+        workflow_approvals = WorkflowApproval.objects.filter(status='pending')
+        now = tz_now()
+        for task in workflow_approvals:
+            approval_timeout_seconds = timedelta(seconds=task.timeout)
+            if task.timeout == 0:
+                continue
+            if (now - task.created) >= approval_timeout_seconds:
+                timeout_message = _("The approval node {name} ({pk}) has expired after {timeout} seconds.").format(
+                    name=task.name, pk=task.pk, timeout=task.timeout
+                )
+                logger.warning(timeout_message)
+                task.timed_out = True
+                task.status = 'failed'
+                task.send_approval_notification('timed_out')
+                task.websocket_emit_status(task.status)
+                task.job_explanation = timeout_message
+                task.save(update_fields=['status', 'job_explanation', 'timed_out'])
 
-        dependent_tasks = dependent_tasks or []
+    @timeit
+    def get_tasks(self):
+        return self.get_running_workflow_jobs()
```

Comment:
Let me volunteer that, before you started on this, `get_running_workflow_jobs` was an atrocious method. Running workflow jobs were already fetched by `get_tasks` and this is one of those duplicate queries that @kdelee has discussed.

With this as a separate class, it could make more sense, but it still gets called in `process_pending_tasks`, which might be because you haven't finished yet. But I wanted to call it out.